pkg/ingester/instance.go

```diff
@@ -381,27 +381,48 @@ func (i *instance) QuerySample(ctx context.Context, req logql.SelectSampleParams
 	return iters, nil
 }
 
-func (i *instance) Label(_ context.Context, req *logproto.LabelRequest) (*logproto.LabelResponse, error) {
-	var labels []string
-	if req.Values {
-		values, err := i.index.LabelValues(req.Name, nil)
-		if err != nil {
-			return nil, err
-		}
-		labels = make([]string, len(values))
-		for i := 0; i < len(values); i++ {
-			labels[i] = values[i]
+func (i *instance) Label(_ context.Context, req *logproto.LabelRequest, matchers ...*labels.Matcher) (*logproto.LabelResponse, error) {
+	if len(matchers) == 0 {
+		var labels []string
+		if req.Values {
+			values, err := i.index.LabelValues(req.Name, nil)
+			if err != nil {
+				return nil, err
+			}
+			labels = make([]string, len(values))
+			for i := 0; i < len(values); i++ {
+				labels[i] = values[i]
+			}
+		} else {
+			names, err := i.index.LabelNames(nil)
+			if err != nil {
+				return nil, err
+			}
+			labels = make([]string, len(names))
+			for i := 0; i < len(names); i++ {
+				labels[i] = names[i]
+			}
 		}
-	} else {
-		names, err := i.index.LabelNames(nil)
-		if err != nil {
-			return nil, err
+		return &logproto.LabelResponse{
+			Values: labels,
+		}, nil
+	}
+
+	ids, _ := i.index.Lookup(matchers, nil)
+	labels := make([]string, 0)
+	for _, streamID := range ids {
+		stream, ok := i.streamsByFP[streamID]
```

Comment:
`i.streamsbyFP` needs to be accessed under the `streamsMtx`. I'd also suggest using the [`forMatchingStreams`](https://github.com/grafana/loki/blob/main/pkg/ingester/instance.go#L497) helper if you're just looking up streams by matchers alone.

Reply:
`forMatchingStreams` is used now. Thanks for the suggestion, it makes the code easier.